### PR TITLE
Added Focal Loss implementation and unit tests

### PR DESCRIPTION
**Pull request by NikoLaFerrari**
This PR introduces an implementation of Focal Loss (`FocalLossType`) to mlpack's neural network loss functions.

`src/mlpack/methods/ann/loss_functions/focal_loss.hpp` & `focal_loss_impl.hpp`:
- Implemented `FocalLossType` template supporting customizable `gamma` (focusing parameter), `alpha` (balancing factor), and `reduction` ('sum' vs 'mean') settings.
- Implemented numerically stable `Forward()` and `Backward()` passes operating directly on raw input logits.
- Added cereal serialization support via `serialize()`.

`src/mlpack/methods/ann/loss_functions/loss_functions.hpp`:
- Included `focal_loss.hpp` in the main loss functions convenience header.

`src/mlpack/tests/ann/loss_functions_test.cpp`:
- Added `FocalLossTest` to verify `Forward()` output, output matrix dimensions, and analytical gradient accuracy against numerical finite-difference gradient checks.

### Testing
Ran the test suite locally:

```bash
./bin/mlpack_test "FocalLossTest"
```

NB: Gemini was used to generate the test.

### CODE REVIEW

**Comment by rcurtin** [comment]
Thanks for the contribution.  Do you have a downstream application of this layer?  I am wondering what the purpose of contributing this back to mlpack is and what the larger need for it is.

**Comment by NikoLaFerrari** [comment]
Hello @rcurtin! Thank you for reviewing.
Tbh, the primary goal was to familiarise myself with mlpack by making a first contribution. That said, the general need for Focal Loss is to combat the struggle of std cross-entropy when working on datasets with class imbalance. It weights down easy examples to keep grad focused on hard examples. However, if adding this adds unecessary maintenance, I understand. Would it be helpful if I added a small benchmark on an imbalanced dataset? Or would you rather have the contributions focused on another area of `mlpack`?

**Comment by andrewfurey21** [comment]
These params could be added to the `SigmoidCrossEntropyErrorType` class, no? Instead of having both.

**Comment by NikoLaFerrari** [comment]
That's a fair point. Mathematically, Focal Loss reduces to Sigmoid Cross-Entropy when $\gamma = 0$ and $\alpha = 0.5$. 

main reason for keeping `FocalLossType` separate was:
1. Avoiding the $(1 - p_t)^\gamma$ exponentiation and element wise matmuls during the passes.
2. PyTorch/TF conventions keep `BCEWithLogitsLoss` and `FocalLoss` as distinct classes to keep standard cross-entropy lightweight.

That said, if it is prefered to keep them into `SigmoidCrossEntropyErrorType`, I'm happy to refactor it!

**Comment by NikoLaFerrari** [comment]
Hello!
I dont understand why it says the code fails Cross-compilation checks. When I check more details, it says it passes all the tests. I am new to this, please do guide me!
